### PR TITLE
fix(session): require explicit repoRoot to prevent silent recording failures

### PR DESCRIPTION
**Pull request by rsnodgrass**
## Summary

Eliminates silent session recording failures caused by `FindSessionFile` deriving `repoRoot` from ambient state (env vars, cwd) which could resolve to `"."` or `""`, silently hashing to a garbage path and losing entire sessions.

- **Interface change**: `FindSessionFile(agentID, since)` → `FindSessionFile(SessionLookup)` where `SessionLookup` carries an explicit, validated `RepoRoot`
- **Defense-in-depth validation** at four layers: `SessionLookup.Validate()`, `validateRepoRootArg()` (CLI pre-flight), `ValidateRepoRoot()` (adapter SDK), and compliance suite (binary-level black-box test)
- **Stop-time hardening**: Uses `RecordingState.WorkspacePath` (persisted at start) as single source of truth — never derives repoRoot from ambient cwd/env at stop time
- **Agent type canonicalization**: `CanonicalAdapterName()` prevents drift ("claude" vs "claude-code") in metadata across all write paths
- **Recovery guidance**: When session file discovery fails at stop time, prints exact `ox agent <id> session import` command and sets doctor marker

```mermaid
flowchart TD
    A[Session Start] -->|persists WorkspacePath| B[RecordingState]
    B -->|hook fires| C{SessionFile found?}
    C -->|no| D[FindSessionFile with explicit repoRoot]
    D --> E{SessionLookup.Validate}
    E -->|repoRoot invalid| F[Error: requires absolute path + .sageox/]
    E -->|valid| G[Adapter binary]
    G --> H{validateRepoRootArg pre-flight}
    H -->|invalid| F
    H -->|valid| I{ValidateRepoRoot SDK}
    I -->|invalid| F
    I -->|valid| J[Search for session JSONL]
    C -->|yes| K[Read entries incrementally]
    
    L[Session Stop] -->|reads WorkspacePath| D
    L -->|file not found| M[Recovery: print import command + doctor marker]
```

## Test plan

- [x] 11 `ValidateRepoRoot` tests (empty, dot, relative, no .sageox, spaces, trailing slash, symlink, file-not-dir, nonexistent, tilde, valid)
- [x] 7 `validateRepoRootArg` edge case tests (last-arg-no-value, duplicate args, equals syntax, mixed args, spaces, nil/empty slice)
- [x] 8 `SessionLookup.Validate` edge case tests (spaces, trailing slash, symlink, file-not-dir, zero-since, optional session ID, nonexistent, tilde)
- [x] 29 `CanonicalAdapterName` subtests covering all known aliases
- [x] 5 one-shot dispatch rejection tests + 1 serve-mode rejection test in `adapterruntime`
- [x] 4 `RecordingState.WorkspacePath` persistence tests (round-trip, legacy compat, spaces, start-sets-path)
- [x] Compliance suite `TestFindSessionBadRepoRoot` for binary-level validation
- [x] `make lint` clean, `make test` passes (1 pre-existing flaky timeout unrelated to changes)

Co-Authored-By: SageOx <ox@sageox.ai>

## Summary by CodeRabbit

* **New Features**
  * Adds session recovery metadata and explicit recovery guidance when session discovery fails.

* **Bug Fixes**
  * Repo-root is now validated and required for session discovery; implicit cwd fallbacks are logged or rejected.
  * Session lookup errors surface earlier with clearer messages; adapter names are normalized for consistent metadata.

* **Tests**
  * Expanded tests for repo-root validation, session discovery, adapter behavior, and canonicalization.

### CODE REVIEW

File: cmd/ox-adapter-aider/hooks.go [hunk 1]
```diff
@@ -150,9 +151,9 @@ func handleUninstallHooks(p adapterprotocol.HookParams) (*adapterprotocol.Uninst
...
 func resolveConventionsPath(repoRoot string) string {
-	if repoRoot != "" {
-		return filepath.Join(repoRoot, "CONVENTIONS.md")
+	if repoRoot == "" {
+		log.Println("WARN: resolveConventionsPath called with empty repoRoot, falling back to cwd")
+		repoRoot, _ = os.Getwd()
 	}
-	cwd, _ := os.Getwd()
-	return filepath.Join(cwd, "CONVENTIONS.md")
+	return filepath.Join(repoRoot, "CONVENTIONS.md")
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox-adapter-aider/hooks.go:158]: Fixed.

File: cmd/ox-adapter-opencode/hooks.go [hunk 2]
```diff
@@ -92,9 +93,9 @@ func resolvePluginPath(repoRoot, scope string) string {
...
 	}
-	if repoRoot != "" {
-		return filepath.Join(repoRoot, openCodeProjectPath, openCodePluginFileName)
+	if repoRoot == "" {
+		log.Println("WARN: resolvePluginPath called with empty repoRoot, falling back to cwd")
+		repoRoot, _ = os.Getwd()
 	}
-	cwd, _ := os.Getwd()
-	return filepath.Join(cwd, openCodeProjectPath, openCodePluginFileName)
+	return filepath.Join(repoRoot, openCodeProjectPath, openCodePluginFileName)
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox-adapter-opencode/hooks.go:100]: Fixed.

File: cmd/ox-adapter-pi/hooks.go [hunk 3]
```diff
@@ -150,9 +151,9 @@ func handleUninstallHooks(p adapterprotocol.HookParams) (*adapterprotocol.Uninst
...
 func resolveAgentsMDPath(repoRoot string) string {
-	if repoRoot != "" {
-		return filepath.Join(repoRoot, "AGENTS.md")
+	if repoRoot == "" {
+		log.Println("WARN: resolveAgentsMDPath called with empty repoRoot, falling back to cwd")
+		repoRoot, _ = os.Getwd()
 	}
-	cwd, _ := os.Getwd()
-	return filepath.Join(cwd, "AGENTS.md")
+	return filepath.Join(repoRoot, "AGENTS.md")
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox-adapter-pi/hooks.go:158]: Fixed.

File: cmd/ox/adapter_test_helpers_test.go [hunk 4]
```diff
@@ -17,7 +17,7 @@ type testClaudeCodeAdapter struct{}
 
 func (a *testClaudeCodeAdapter) Name() string  { return "claude-code" }
 func (a *testClaudeCodeAdapter) Detect() bool  { return false }
-func (a *testClaudeCodeAdapter) FindSessionFile(_ string, _ time.Time) (string, error) {
+func (a *testClaudeCodeAdapter) FindSessionFile(_ adapters.SessionLookup) (string, error) {
 	return "", adapters.ErrSessionNotFound
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/adapter_test_helpers_test.go:21]: Fixed.

File: internal/adapterprotocol/compliance/suite.go [hunk 5]
```diff
@@ -117,6 +119,48 @@ func (s *Suite) TestDetect(t *testing.T) {
...
+	}
+
+	// non-zero exit with stderr is also acceptable
+	if len(stderr) > 0 {
+		return
+	}
+
+	// non-zero exit with no output at all is still acceptable (it's an error)
+	_ = err
+}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/adapterprotocol/compliance/suite.go:162]: Fixed.

File: internal/daemon/daemon.go [hunk 6]
```diff
@@ -320,10 +324,14 @@ func (d *Daemon) Start() error {
 
 	// register in daemon registry for multi-daemon support
 	// use ProjectRoot (the actual workspace), not LedgerPath
+	// NOTE: must happen BEFORE StabilizeCWD() which changes cwd to $HOME
 	workspacePath := d.config.ProjectRoot
 	if workspacePath == "" {
+		d.logger.Warn("daemon config missing ProjectRoot, falling back to cwd")
 		workspacePath, _ = os.Getwd()
 	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/daemon/daemon.go:332]: Fixed.

File: internal/session/adapters/adapter.go [hunk 7]
```diff
@@ -25,6 +27,44 @@ var (
...
+// RepoRoot must be absolute and contain a .sageox/ directory.
+func (sl SessionLookup) Validate() error {
+	if sl.RepoRoot == "" || sl.RepoRoot == "." {
+		return fmt.Errorf("repoRoot must be absolute, got %q", sl.RepoRoot)
+	}
+	if !filepath.IsAbs(sl.RepoRoot) {
+		return fmt.Errorf("repoRoot %q is not absolute", sl.RepoRoot)
+	}
+	if info, err := os.Stat(filepath.Join(sl.RepoRoot, ".sageox")); err != nil || !info.IsDir() {
+		return fmt.Errorf("repoRoot %q has no .sageox/ directory", sl.RepoRoot)
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/session/adapters/adapter.go]: Fixed.

File: internal/session/adapters/generic_jsonl_test.go [hunk 8]
```diff
@@ -23,7 +23,7 @@ func TestGenericJSONLAdapter_Detect(t *testing.T) {
 
 func TestGenericJSONLAdapter_FindSessionFile(t *testing.T) {
 	adapter := &GenericJSONLAdapter{}
-	_, err := adapter.FindSessionFile("agent-123", time.Now())
+	_, err := adapter.FindSessionFile(SessionLookup{RepoRoot: "/tmp", AgentID: "agent-123", Since: time.Now()})
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/session/adapters/generic_jsonl_test.go:26]: Fixed.

File: internal/session/history_store.go [hunk 9]
```diff
@@ -85,6 +86,7 @@ func GetHistoryStoragePath(agentID string, activeRecording bool) string {
 	// try to find active recording state
 	cwd, err := os.Getwd()
 	if err != nil {
+		slog.Warn("GetHistoryStoragePath: os.Getwd failed, using '.'", "error", err)
 		cwd = "."
 	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/session/history_store.go:91]: Fixed.

File: internal/session/recording_state_test.go [hunk 10]
```diff
@@ -516,6 +516,97 @@ func TestLoadAllRecordingStates_MixedValidAndCorrupt(t *testing.T) {
...
+// Failure prevented: legacy .recording.json without workspace_path causes nil panic.
+func TestRecordingState_WorkspacePath_EmptyForLegacy(t *testing.T) {
+	cacheDir := t.TempDir()
+	projectRoot, sessionsBase := setupRecordingTestWithSessionsBase(t, cacheDir)
+	sessionPath := filepath.Join(sessionsBase, "2026-01-01T10-00-user-OxLgcy")
+	require.NoError(t, os.MkdirAll(sessionPath, 0o755))
+
+	// simulate old recording state JSON without workspace_path field
+	legacyJSON := `{"agent_id":"OxLgcy","started_at":"2026-01-01T10:00:00Z","adapter_name":"claude-code","session_path":"` + sessionPath + `"}`
+	require.NoError(t, os.WriteFile(filepath.Join(sessionPath, recordingFile), []byte(legacyJSON), 0o600))
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/session/recording_state_test.go:558]: Fixed.

File: pkg/adapterruntime/runtime.go [hunk 11]
```diff
@@ -16,13 +16,31 @@ import (
...
+	}
+	if !filepath.IsAbs(root) {
+		return fmt.Errorf("repo-root %q is not absolute", root)
+	}
+	info, err := os.Stat(filepath.Join(root, ".sageox"))
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("repo-root %q has no .sageox/ directory", root)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply pkg/adapterruntime/runtime.go:52]: Fixed.

File: cmd/ox/agent_session.go [hunk 12]
```diff
@@ -530,6 +594,119 @@ func runAgentSessionStop(inst *agentinstance.Instance) error {
...
+	if len(candidates) < limit {
+		limit = len(candidates)
+	}
+	for _, c := range candidates[:limit] {
+		if fileContainsString(c.path, agentID) {
+			return c.path
+		}
+	}
+	return ""
+}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/agent_session.go:695]: Fixed.
